### DISCUSSION TITLE
Verbose support for tryConsume with a wait duration

### DISCUSSION THREAD

**Discussion by HaloFour** [vladimir-bukhtoyarov/bucket4j#482] [Ideas]
I'm using the `VerboseBucket` interface to call `tryConsume` in my Spring service for rate limiting, so that I can accurately report the remaining tokens in metrics.  I am interested in using the overload of `tryConsume` that accepts a wait time so that I can introduce a short delay when out of tokens but I noticed that these overloads aren't available through `VerboseBucket`.  I'd like to be able to report both the remaining tokens, if available, as well as the delay time, if necessary.  There is the verbose implementation of `estimateAbilityToConsume`, but there isn't a thread-safe way to report that information _and_ consume the tokens.

Is there any interest in adding this API?  I'd be happy to contribute it.

_Edit_:  I imagine it might be more appropriate to expand `tryConsumeAndReturnRemaining` so that it supports a wait time, given it already seems to handle the rest and does have a "verbose" version as well?

**Comment by vladimir-bukhtoyarov** [comment]
Hello @HaloFour 

In the scope of #117 it was especially decided to not provide `bucket.asBlocking().asVerbose()` API and its friend `bucket.asScheduler().asVerbose()` was not added as well. Because returned Verbose result can become significantly inactual after sleeping(or postponed inside sheduler-service).  If we asquire snaphsot [at this point](https://github.com/bucket4j/bucket4j/blob/master/bucket4j-core/src/main/java/io/github/bucket4j/AbstractBucket.java#L241) then it can become [significantly obsolete](https://github.com/bucket4j/bucket4j/blob/master/bucket4j-core/src/main/java/io/github/bucket4j/AbstractBucket.java#L251) because of `blockingStrategy.parkUninterruptibly(nanosToSleep)`. 

In general it is not a big deal to implement the both `bucket.asBlocking().asVerbose()` and `bucket.asScheduler().asVerbose()` but I do not want to do this, because providing obsolete info will be a cause of issues in Github bugtracker. So, at least it needs to find sppropriate words  for Javadoc that point attention that snaphot is actual to moment of time that prior to going to be blocked. But I am worry about that mostly nobody reads the Javadoc.

**Comment by vladimir-bukhtoyarov** [comment]
After thinking more I decided to implement such feature #483. I have changed my mind because released that, the main purpuse of VerboseResult is providing the answer why any method returns concrete result. Providing most fresh snapshot is not a goal, so it nothing wrong to return Snapshot that can be obsolete at return time.

**Comment by vladimir-bukhtoyarov** [comment]
@HaloFour 

Since `8.12.0` it will be possible to write code like this:
```java
Duration maxWait = Duration.ofMillis(10);
VerboseResult<Boolean> tryConsumeResponse = bucket.asBlocking().asVerbose().tryConsume(1, maxWait);
boolean wasConsumed = tryConsumeResponse.getValue();
if (wasConsumed) {
   // do something
} else {
  // do something else
}
```

**Reply by HaloFour** [reply]
Thank you!  I will check this out!
